### ISSUE TITLE
Error handling on thread level

### ISSUE THREAD

**Issue by dkropachev** [scylladb/cassandra-stress#29]
When the test starts `c-s` spins threads to generate a load, you can control the number of threads by `-rate threads=X`.
Unless `-errors ignore` is provided, the thread is going to be killed after reaching a number of errors.

The problem is that `c-s` does not stop the test when thread is killed, even if all of them are killed.

Recently we have merged https://github.com/scylladb/cassandra-stress/pull/26, which has changed `c-s` behavior to the following:
1. If `-errors fail-fast` is provided it would fail whole run after first thread is killed
2. Otherwise it would keep waiting, unless all the threads are killed.

Now, in my book there are only three correct behaviors in regards to the threads handling:
1. Fail the test when the first thread is killed. Currently works only if `-errors fail-fast`
2. Not to kill a thread when error is encountered. keep it running. Currently works only if `-errors ignore` is provided
3. Fail all thread at once when they reach error state, but keep them running untill that point. Not implemented currently

Anything in between I see as unwanted behavior that can lead to unexpected, unpredictable or inconsitant results.
On this issue I want to start discussion to decide on what is the `correct` behavior` and implement it eventually.

Migrated to Jira: [QATOOLS-17](https://scylladb.atlassian.net/browse/QATOOLS-17)

[QATOOLS-17]: https://scylladb.atlassian.net/browse/QATOOLS-17?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by dkropachev** [comment]
@CodeLieutenant , @Bouncheck , @fruch, please take a look

**Comment by CodeLieutenant** [comment]
> 3\. Fail all thread at once when they reach error state, but keep them running untill that point. Not implemented currently

I don't really understand this point, this looks like to me a `fail-fast`. Kill all other threads when on thread reaches error state. Am I getting it correctly, or run and kill threads one by one, as they encounter errors.

We also need to decide what should be the default behaviour when `-errors` is not passed in, and whats the current state of the system

**Comment by dkropachev** [comment]
> > 3. Fail all thread at once when they reach error state, but keep them running untill that point. Not implemented currently
> 
> I don't really understand this point, this looks like to me a `fail-fast`. Kill all other threads when on thread reaches error state. Am I getting it correctly, or run and kill threads one by one, as they encounter errors.
> 
> We also need to decide what should be the default behaviour when `-errors` is not passed in, and whats the current state of the system

It is like `fail-fast` but it does no fail fast, it ensures that all threads are in failed state before failing stress command, which suppose to give more stability.

**Comment by CodeLieutenant** [comment]
> It is like `fail-fast` but it does no fail fast, it ensures that all threads are in failed state before failing stress command, which suppose to give more stability.

So, fail threads one by one with no restart?

**Comment by dkropachev** [comment]
> > It is like `fail-fast` but it does no fail fast, it ensures that all threads are in failed state before failing stress command, which suppose to give more stability.
> 
> So, fail threads one by one with no restart?

No, fail them all at once, only when all threads can't execute a single query, otherwise all threads continue trying executing queries.
